[Form] Added article for custom choice fields (#13490)
symfony/symfony-docs: form/create_custom_choice_type.rst

        ->add('category', CategoryChoiceType::class, [
            // ... some inherited or custom options for that type
        ])
        // ... add more fields
    ;

In such case you can create a type in many ways as explained below.

Creating a Type With Static Custom Choices
------------------------------------------

To create a custom choice type when choices are static, you can do the
following::

    // src/Form/Type/CategoryChoiceType.php
    namespace App\Form\Type;

    use App\Domain\Model;
    use Symfony\Component\Form\AbstractType;
    use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
    use Symfony\Component\OptionsResolver\OptionsResolver;

    class CategoryChoiceType extends AbstractType
    {
        /**
         * {@inheritdoc}
         */
        public function getParent()
        {
            // inherits all options, form and view configuration
            // to create expanded or multiple choice lists
            return ChoiceType::class;
        }

        /**
         * {@inheritdoc}
         */
        public function configureOptions(OptionsResolver $resolver)
        {
            $resolver
                // override the "choices" option

>                 ->setDefault('choices', Model::getCategories())

Reviewer: Above this line:

```
// Use whatever way you want to get the choices - Mode::getCategories() is just an example
```

Author: Done, thanks!

    defined.
    Choices can be grouped with keys as group name and nested iterable choices
    in alternative to the :ref:`group_by <reference-form-group-by>` option.

`doLoadChoicesForValues(array $values): array`

    Optional, to improve performance this method is called when the data is
    submitted. You can then load the choices partially, by using the submitted
    values passed as only argument.
    The list is fully loaded by default.

`doLoadValuesForChoices(array $choices): array`

    Optional, in alternative to the
    `choice_value <reference-form-choice-value>` option.
    You can implement this method to return the string values partially, the
    initial choices are passed as only argument.
    The list is full loaded by default unless the ``choice_value`` option is
    defined.

Then you need to update the previous example to use the new loader instead::

    // src/Form/Type/CategoryChoiceType.php;

    // ... same as before
    use App\Api\AcmeApi;
    use App\Form\ChoiceList\AcmeCategoryLoader;

    class CategoryChoiceType extends AbstractType
    {
        // using the default configuration, the type is a service
        // so the api will be autowired

>         public AcmeApi $api;

Reviewer: At this time, using `public` properties for autowiring is not the way we're showing in the docs. And it can't be (everything else aside) until we're writing docs for a Symfony version that requires at least PHP 7.4

Author: I wanted to avoid the noise of private + construct since I target master, but you're right. I've changed it.

.. index::
   single: Form; Custom choice type

How to Create a Custom Choice Field Type
========================================

Symfony :doc:`ChoiceType </reference/forms/types/choice>` is a very useful type
that deals with a list of selected options.
The Form component already provides many different choice types, like the
intl types (:doc:`LanguageType </reference/forms/types/language>`, ...) and the
:doc:`EntityType </reference/forms/types/entity>` which loads the choices from
a set of Doctrine entities.

It's also common to want to re-use the same list of choices for different fields.
Creating a custom "choice" field is a great solution - something like::

    use App\Form\Type\CategoryChoiceType;

    // ... from any type
    $builder
        ->add('category', CategoryChoiceType::class, [
            // ... some inherited or custom options for that type
        ])

>         // ... add more fields

Reviewer 2: I don't think each collapse comment (``...``) needs extra information. I would remove the words in this one (and maybe on line 19 as well). In the rest of the docs, we're only adding the comment for places where the user must do something (as an emphasize), so this can confuse people.

Author: I've removed as I agree it's not useful, but kept the one line 19 because this is the only context hint in that example code.

    ;

Creating a Type With Static Custom Choices
------------------------------------------

To create a custom choice type when choices are static, you can do the
following::

    // src/Form/Type/CategoryChoiceType.php
    namespace App\Form\Type;

    use App\Domain\Model;
    use Symfony\Component\Form\AbstractType;
    use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
    use Symfony\Component\OptionsResolver\OptionsResolver;

    class CategoryChoiceType extends AbstractType
    {
        /**
         * {@inheritdoc}
         */
        public function getParent()
        {
            // inherits all options, form and view configuration
            // to create expanded or multiple choice lists
            return ChoiceType::class;
        }

        /**
         * {@inheritdoc}
         */
        public function configureOptions(OptionsResolver $resolver)
        {
            $resolver

>                 // Using whatever way you want to get the choices - Mode::getCategories() is just an example

Reviewer 2: ```suggestion
                // Use whatever way you want to get the choices - Mode::getCategories() is just an example
```

    }

.. caution::

     The ``getParent()`` method is used instead of ``extends``.
     This allows the type to inherit from both ``FormType`` and ``ChoiceType``.

Loading Lazily Static Custom Choices
------------------------------------

Sometimes, the callable to define the ``choices`` option can be a heavy process
that could be prevented when the submitted data is optional and empty.
Sometimes it can depend on other options.

The solution is to load the choices lazily using the ``choice_loader`` option,
which accepts a callback::

    use Symfony\Component\Form\ChoiceList\ChoiceList;
    use Symfony\Component\OptionsResolver\Options;

    // use this option instead of the "choices" option
    ->setDefault('choice_loader', ChoiceList::lazy($this, static function() {
        return Model::getCategories();
    }))

    // or if it depends on other options
    ->setDefault('some_option', 'some_default')
    ->setDefault('choice_loader', static function (Options $options) {
        $someOption = $options['some_option'];

        return ChoiceList::lazy($this, static function() use ($someOption) {
            return Model::getCategories($someOption);
        }, $someOption);

>     }))

Reviewer 2: I would always make sure that our code examples are syntax correct (because it's often easier to read - you know how to call the method - and the Pygments code highlighter is sometimes very strict in these things - making the example completely read if there is something wrong).

So in this code example,  please add `$resolver` to each call

Author: Fixed thanks!

.. note::

>     We use the ``ChoiceList::lazy()`` method to create a cached

Reviewer 2: Please don't use first-person in the docs (also in the rest of the article): "Use the `ChoiceList::lazy()`method to create a cached ..."

Author: Changed for `The ``ChoiceList::lazy()`` method creates a cached...`

    :class:`Symfony\\Component\\Form\\ChoiceList\\Loader\\CallbackChoiceLoader`
    object. The first argument ``$this`` is the type configuring the form, and
    a third argument ``$vary`` can be used as array to pass any value that
    makes the loaded choices different.

Creating a Type With Dynamic Choices
------------------------------------

When loading choices is complex, a callback is not enough and a "real" service
is needed. Fortunately, the Form component provides a
:class:`Symfony\\Component\\Form\\ChoiceList\\Loader\\ChoiceLoaderInterface`.
You can pass any instance to the ``choice_loader`` option to handle things
any way you need. For example, you could leverage this new power to load
categories from an HTTP API. The easiest way is to extend the
:class:`Symfony\\Component\\Form\\ChoiceList\\Loader\\AbstractChoiceLoader`
class, which already implements the interface and avoids triggering your logic

> when it is not needed (i.e when the form is submitted empty and valid).

Reviewer 2: ```suggestion
when it is not needed (e.g. when the form is submitted empty and valid).
```

"e.g." is "for example", "i.e." is "in other words".

Author: I meant it, but after thinking about this again, there is another case actually. When defining the `choice_value` option, `doLoadValuesForChoices` won't be called. I've removed that note before opening the PR, but maybe I should add it again?

This could look like this::

    // src/Form/ChoiceList/AcmeCategoryLoader.php.
    namespace App\Form\ChoiceList;

    use App\Api\AcmeApi;
    use Symfony\Component\Form\ChoiceList\Loader\AbstractChoiceLoader;

    class AcmeCategoryLoader extends AbstractChoiceLoader;
    {

>         // This must be passed by the type

Reviewer 2: ```suggestion
        // this must be passed by the type
```

I figured my review already had enough comments to also be very nitpicky about these little changes 🙈

Author: Fixed!

        // this loader won't be registered as service
        private $api;
        // define more options if needed
        private $someOption;

        public function __construct(AcmeApi $api, string $someOption)
        {
            $this->api = $api;
            $this->someOption = $someOption;
        }

        protected function loadChoices(): iterable
        {
            return $this->api->loadCategories($this->someOption));
        }

        protected function doLoadChoicesForValues(array $values): array
        {
            return $this->api->loadCategoriesForNames($values, $this->someOption);
        }

        protected function doLoadValuesForChoices(array $choices): array
        {
            $values = [];

>             // compute string values that must be submitted

Reviewer 2: ```suggestion
            // ... compute string values that must be submitted
```

            return $values;
        }
    }

Here we implement three protected methods:

> `loadChoices(): iterable`

Reviewer 2: ```suggestion
``loadChoices(): iterable``
```

>     This method is abstract and is the only one that needs to be implemented,
>     it is called when the list is fully loaded (i.e when rendering the view).

Reviewer 2: ```suggestion
    This method is abstract and is the only one that needs to be implemented.
    It is called when the list is fully loaded (i.e when rendering the view).
```

    It must return an array or a traversable object, keys are default labels
    unless the :ref:`choice_label <reference-form-choice-label>` option is
    defined.
    Choices can be grouped with keys as group name and nested iterable choices
    in alternative to the :ref:`group_by <reference-form-group-by>` option.

> `doLoadChoicesForValues(array $values): array`

Reviewer 2: ```suggestion
``doLoadChoicesForValues(array $values): array``
```

    Optional, to improve performance this method is called when the data is
    submitted. You can then load the choices partially, by using the submitted
    values passed as only argument.
    The list is fully loaded by default.

> `doLoadValuesForChoices(array $choices): array`

Reviewer 2: ```suggestion
``doLoadValuesForChoices(array $choices): array``
```

>     Optional, in alternative to the

Reviewer 2: ```suggestion
    Optional, as alternative to the
```

(not very sure about this one)

Author: I've fixed, maybe @weaverryan missed it or I will revert. You made me doubt :D.

>     `choice_value <reference-form-choice-value>` option.

Reviewer 2: ```suggestion
    :ref:`choice_value <reference-form-choice-value>` option.
```

    You can implement this method to return the string values partially, the
    initial choices are passed as only argument.
    The list is fully loaded by default unless the ``choice_value`` option is
    defined.

> Then you need to update the previous example to use the new loader instead::

Reviewer 2: Maybe instead of "the previous example" use "the form type"?

Author: Agreed, updated, thanks!

    // src/Form/Type/CategoryChoiceType.php;

>     // ... same as before

Reviewer 2: ```suggestion
    // ...
```

These is one such place where I don't think the words add anything (same below)

Author: I've removed this one and kept the other below as it's read "same as before but..."

    use App\Api\AcmeApi;
    use App\Form\ChoiceList\AcmeCategoryLoader;

    class CategoryChoiceType extends AbstractType
    {
        // using the default configuration, the type is a service
        // so the api will be autowired
        private $api;

        public function __construct(AcmeApi $api)
        {
            $this->api = $api;
        }

        // ... same as before

        public function configureOptions(OptionsResolver $resolver)
        {
            $resolver
                // ... same as before
                // but use the custom loader instead

>                 ->setDefault('choice_loader', static function(Options $options) {

Reviewer 2: Do we really want `static function(...) {}` in the code? I know it's how it's done in the core code, but I think it adds complexity without benefit.
Docs are imho somewhere in the middle between not providing bad practices, but also not providing the perfect practice to keep things simple enough to be understand (and does static even work in this case, as we're using `$this` and `$this->api` in the closure body?)

Author: I've added them afterwards, and introduced that mistake, whereas I usually do this with caution... Well I wouldn't check and update all the docs, but I would be in favor to consider this a good practice and use it when relevant, e.g showing example userland code that may be copy/pasted like here and not using `$this`.

the same as before, with some minor differences::

    // src/Form/Type/CategoryChoiceType.php;

    // ...

    use App\Entity\AcmeCategory;
    use Symfony\Bridge\Doctrine\Form\Type\EntityType;

    class CategoryChoiceType extends AbstractType
    {
        public function getParent()
        {
            return EntityType::class;
        }

        public function configureOptions(OptionsResolver $resolver)
        {
            $resolver
                // can now override options from both entity and choice types
                ->setDefault('class', AcmeCategory::class)

                // you can also customize the "query_builder" option
                ->setDefault('some_option', 'some_default')
                ->setDefault('query_builder', static function(Options $options) {
                    $someOption = $options['some_option'];

                    return static function (AcmeCategoryRepository $repository) use ($someOption) {
                        return $repository->createQueryBuilderWithSomeOption($someOption);
                    };
                })
            ;
        }
    }

Customize Templates
-------------------

> This section is already covered in
> :doc:`create a custom field type </form/create_custom_field_type>`.

Reviewer 2: ```suggestion
Read ":doc:`/form/create_custom_field_type`" on how to customize
the form themes for your new choice field type.
```

                // ... override more choice options or define new ones
            ;
        }
    }

.. caution::

     The ``getParent()`` method is used instead of ``extends``.
     This allows the type to inherit from both ``FormType`` and ``ChoiceType``.

Loading Lazily Static Custom Choices
------------------------------------

Sometimes, the callable to define the ``choices`` option can be a heavy process
that could be prevented when the submitted data is optional and empty.
Sometimes it can depend on other options.

The solution is to load the choices lazily using the ``choice_loader`` option,
which accepts a callback::

    use Symfony\Component\Form\ChoiceList\ChoiceList;
    use Symfony\Component\OptionsResolver\Options;

    $resolver
        // use this option instead of the "choices" option
        ->setDefault('choice_loader', ChoiceList::lazy($this, static function() {
            return Model::getCategories();
        }))

        // or if it depends on other options
        ->setDefault('some_option', 'some_default')
        ->setDefault('choice_loader', static function(Options $options) {
            $someOption = $options['some_option'];

>             return ChoiceList::lazy($this, static function() use ($someOption) {

Author: Good catch, fixed. Thanks!

    defined.
    Choices can be grouped with keys as group name and nested iterable choices
    in alternative to the :ref:`group_by <reference-form-group-by>` option.

``doLoadChoicesForValues(array $values): array``

    Optional, to improve performance this method is called when the data is
    submitted. You can then load the choices partially, by using the submitted
    values passed as only argument.
    The list is fully loaded by default.

``doLoadValuesForChoices(array $choices): array``

    Optional, as alternative to the
    :ref:`choice_value <reference-form-choice-value>` option.
    You can implement this method to return the string values partially, the
    initial choices are passed as only argument.
    The list is fully loaded by default unless the ``choice_value`` option is
    defined.

Then you need to update the form type to use the new loader instead::

    // src/Form/Type/CategoryChoiceType.php;

    // ... same as before
    use App\Api\AcmeApi;
    use App\Form\ChoiceList\AcmeCategoryLoader;

    class CategoryChoiceType extends AbstractType
    {
        // using the default configuration, the type is a service

>         // so the api will be autowired

Reviewer 3: not sure this is needed on this page - or it would be on all pages?

Author: Maybe, it's sometimes good to have context as ppl can use the component without the framework. And here it adds consistency with this other comment https://github.com/symfony/symfony-docs/pull/13490/files#diff-fa7040ab583bc30dd0ef47caa62a37f3R134
